File: src/core/jev.js
import { ROUTES } from './engines.js';

/**
 * Jev v0: bounded route judgement.
 * This deterministic fallback keeps the UI useful before an external Jev
 * service is configured.
 */
export function judgeRoutes({ instrument, analysis, notes }) {
  const voicedRatio = analysis.totalFrames ? analysis.voicedFrames / analysis.totalFrames : 0;
  const monophonicFriendly = ['fiddle', 'bass', 'vocals'].includes(instrument);
  const physicalBoost = instrument === 'fiddle' ? 4 : monophonicFriendly ? 2 : 0;
  const directBoost = ['guitar', 'piano', 'drums', 'unknown'].includes(instrument) ? 2 : 0;
  const transcriptionEvidence = notes.length >= 3 ? 2 : notes.length ? 1 : -2;
  const pitchEvidence = voicedRatio > 0.25 ? 2 : voicedRatio > 0.08 ? 1 : -1;

  const scored = [
    { ...ROUTES.physical, score: 7 + physicalBoost + transcriptionEvidence + pitchEvidence },
    { ...ROUTES.ddsp, score: 7 + (monophonicFriendly ? 2 : 0) + pitchEvidence },
    { ...ROUTES.direct, score: 7 + directBoost + (analysis.crestFactor < 4 ? 1 : 0) },
  ].sort((a, b) => b.score - a.score);

  return scored.map((route, index) => ({
    ...route,
    rank: index + 1,
    recommended: index === 0,
    reason: routeReason(route.id, { instrument, voicedRatio })
  }));
}

function routeReason(route, context) {
  if (route === 'physical' && context.instrument === 'fiddle') {
    return 'Violin-specific transcription plus a physical renderer matches this stem and can replace the generated waveform entirely.';
  }
  if (route === 'ddsp') {
    return context.voicedRatio > 0.2
      ? 'The stem exposes enough stable pitch evidence for continuous F0/dynamics reconstruction.'
      : 'Useful as a structured benchmark, but pitch evidence is currently limited.';
  }
  return 'Direct transfer preserves the source performance with the least interpretation and is a strong A/B baseline.';
}

File: backend/fidelis_backend/harness/jev.py
from __future__ import annotations

from typing import Any


def judge_routes(*, instrument: str, performance: dict[str, Any] | None, capabilities: dict[str, dict[str, Any]]) -> dict[str, Any]:
    note_count = len((performance or {}).get('performance') or [])
    voiced_ratio = float(((performance or {}).get('analysis') or {}).get('voicedRatio') or 0)

    routes = [
        {
            'id': 'physical',
            'label': 'Performance → physical renderer',
            'stack': ['fidelis-native-performance', 'stradi', 'basic-pitch', 'instrudio-native', 'instrudio', 'fidelis-physical-violin'],
            'score': 12 if instrument in {'fiddle', 'violin', 'strings'} else 5,
            'reason': 'Rebuild a new waveform from recovered performance and a physical/sample renderer.',
        },
        {
            'id': 'ddsp',
            'label': 'Structured DDSP reconstruction',
            'stack': ['ddsp'],
            'score': 9 + (2 if voiced_ratio > .2 else 0),
            'reason': 'Preserve continuous pitch/dynamics while replacing the source timbre through structured synthesis.',
        },
        {
            'id': 'direct',
            'label': 'Direct timbre transfer',
            'stack': ['rave', 'brave', 'sony-diffusion', 'wavetransfer'],
            'score': 8 if instrument in {'guitar', 'piano', 'drums', 'unknown'} else 7,
            'reason': 'Preserve source phrasing directly with the least symbolic interpretation.',
        },
    ]
    if note_count >= 3:
        routes[0]['score'] += 2
        routes[1]['score'] += 1

    for route in routes:
        statuses = [(capabilities.get(adapter) or {}).get('status', 'unavailable') for adapter in route['stack']]
        route['adapterStatuses'] = dict(zip(route['stack'], statuses))
        # A route is executable if at least one direct-transfer engine is ready, or all mandatory structured components are ready.
        if route['id'] == 'direct':
            route['executable'] = any(status == 'ready' for status in statuses)
        elif route['id'] == 'physical':
            transcriber_ready = any((capabilities.get(x) or {}).get('status') == 'ready' for x in ('stradi', 'basic-pitch', 'fidelis-native-performance'))
            violin_class = instrument in {'fiddle', 'violin', 'strings'}
            renderer_ready = violin_class and any((capabilities.get(x) or {}).get('status') == 'ready' for x in ('instrudio-native', 'instrudio', 'fidelis-physical-violin'))
            route['executable'] = transcriber_ready and renderer_ready
        else:
            route['executable'] = all(status == 'ready' for status in statuses)

    routes.sort(key=lambda item: item['score'], reverse=True)
    for i, route in enumerate(routes, 1):
        route['rank'] = i

    executable = [route for route in routes if route['executable']]
    fallback = {
        'id': 'reference',
        'label': 'Fidelis reference renderer',
        'adapterId': 'fidelis-reference-synth',
        'executable': (capabilities.get('fidelis-reference-synth') or {}).get('status') == 'ready',
        'reason': 'Pipeline-validation fallback only; it does not represent the target fidelity path.',
    }
    return {
        'schema': 'fidelis.jev-route-decision.v0.1',
        'instrument': instrument,
        'desiredRoutes': routes,
        'preferredRouteId': routes[0]['id'] if routes else None,
        'executableRouteId': executable[0]['id'] if executable else ('reference' if fallback['executable'] else None),
        'fallback': fallback,
    }
